feat(ros): additional package maps

daenny opened the pull request:
This allows for an additional key in the `package.build.config`:
```toml
extra-package-mappings = ['path-to-file.yaml']
```
This will add/overwrite the mappings from `robostack` e.g. if you host your own packages. 

An additional (later) option would be to let packages allow to install the mapping file inside the build environment. 

I also added a `test` enviroment to the `pixi.toml` such that it is easy to develop and test within pycharm. 

You can now open a pycharm session inside the `pixi-build-ros` `test` environment and select the correct interpreter. Then you can run the `pytest` within `pycharm`.

baszalmstra commented on backends/pixi-build-ros/src/pixi_build_ros/ros_generator.py:
```diff
@@ -19,12 +21,28 @@
+    # Ensure it's an absolute path
+    if not value.is_absolute():
+        # Convert to absolute path relative to the current working directory
+        return Path.cwd() / value
```
This might be dangerous because there are no guarantees that the cwd is set to the root of the package. Its better to rely on the manifest path as the root.

baszalmstra replied:
I dont think that variable is set when used as a backend. The only reliable way forward is to get the path from the manifest_path. Relative paths should always be relative to that file (or its parent dir actually).

baszalmstra commented on backends/pixi-build-ros/src/pixi_build_ros/ros_generator.py:
```diff
@@ -38,21 +56,32 @@ class ROSBackendConfig:
     distro: Optional[str] = None
 
+    # Extra package mappings to use in the build
+    extra_package_mappings: List[Path] = []
```
We use kebab-case in most places, this deviates from that.

daenny replied:
Changed!

baszalmstra reviewed (commented):
Partial review, love the functionality

daenny commented:
Some additions:

* I renamed `robostack` to `conda` in the `packagemap`  that feels more descriptive. WDYT?
* I also added the option to define `rosdistro` packages in the package map, which are resolved using the same rules as if it is defined in the `rosdistro`.

baszalmstra reviewed (commented):
Im not sure about the robostacknto conda rename. It makes it slightly harder to migrate from the robostack based file. Can we instead allow both keys and not rename the keys in the bundled file? 

I would also prefer to create seperate PRs for these changes because this PR is already becoming very large.

nichmor commented on backends/pixi-build-ros/tests/data/other_package_map.yaml:
```diff
@@ -0,0 +1,6 @@
+new_package:
+  conda: [new-package]
+ros_package:
+  ros: [ros_package, ros_package_msgs]
```
I think we need an empty space here

daenny replied:
I added new lines

daenny commented on backends/pixi-build-ros/src/pixi_build_ros/ros_generator.py:
```diff
@@ -19,40 +22,76 @@
+)
+
+
+def _parse_str_or_path(value: str | Path, manifest_root: Path) -> Path:
```
Done

daenny commented on backends/pixi-build-ros/src/pixi_build_ros/ros_generator.py:
```diff
@@ -19,40 +22,76 @@
-    debug_dir: Optional[Path] = None
+    debug_dir: Optional[Path] = pydantic.Field(default=None, alias="debug-dir")
     # Extra input globs to include in the build hash
     extra_input_globs: Optional[List[str]] = None
```
Done

daenny commented on backends/pixi-build-ros/src/pixi_build_ros/ros_generator.py:
```diff
@@ -19,40 +22,76 @@
+        if info.context and "manifest_root" in info.context:
+            base_path = Path(info.context["manifest_root"])
+
+        if input_value is not None:
```
Done

nichmor commented:
hey! I've tested the PR, works as expected!
The only thing that I would like to ask is if I set the new field in the navigator package ( from [ros_work](https://github.com/ruben-arts/ros_workspace) ,  like this:
<img width="1169" height="363" alt="image" src="https://github.com/user-attachments/assets/4674de61-9c34-4ffb-ae0f-5b64343284c0" />


and will try to build it like this:
`PIXI_BUILD_BACKEND_OVERRIDE_ALL=1 pixi build`
It can still build it without errors  ( I would expect an error because path-to-file.yaml doesn't exist ) .

Is it expected?
If yes, we can merge it

nichmor commented:
> @nichmor thanks for the review. I had that as intended behaviour, but indeed it might be confusing. Now I throw an value error. I also fixed the double symlink in the `noetic` build script.

thanks! let me retry it, and I think we can merge it after that

nichmor commented:
works as expected, thanks a lot!
